File: Engineering Concepts/Design Pattern and Principles/Exercise1/Logger.java
public class Logger {

    // 1. Private static instance (single object)
    private static Logger instance;

    // 2. Private constructor (prevents object creation)
    private Logger() {
        System.out.println("Logger instance created");
    }

    // 3. Public method to provide access to the single instance
    public static Logger getInstance() {
        if (instance == null) {
            instance = new Logger(); // lazy initialization
        }
        return instance;
    }

    // Sample method to test logging
    public void log(String message) {
        System.out.println("LOG: " + message);
    }
}

File: Week one handson/Design pattern and principle/SingletonPatternExample/Logger.cs
using System;

namespace SingletonPatternExample
{
    public sealed class Logger
    {
        private static Logger? instance;
        private static readonly object lockObject = new object();

        private Logger()
        {
            Console.WriteLine("Logger Instance Created");
        }

        public static Logger GetInstance()
        {
            lock (lockObject)
            {
                if (instance == null)
                {
                    instance = new Logger();
                }
            }

            return instance;
        }

        public void Log(string message)
        {
            Console.WriteLine("Log: " + message);
        }
    }
}